### PR TITLE
Update the documentation howto index page and group docs into 3 logical sections

### PR DESCRIPTION
**Pull request by willingc**
During PyCon discussions, a number of authors and educators mentioned simplifying the display of information on the HOWTO landing page. This PR groups the HOWTO documents into 3 sections:
- General
- Advanced development
- Debugging and profiling

This PR should help improve docs usability. The sections are sorted alphabetically roughly by the module or important topic.

**Screenshot below**

Screenshot 2024-05-21 at 5 24 26 PM

----
📚 Documentation preview 📚: https://cpython-previews--119366.org.readthedocs.build/en/119366/howto/index.html

### CODE REVIEW

File: Doc/howto/index.rst [hunk 1]
```diff
@@ -35,3 +33,31 @@ Currently, the HOWTOs are:
...
+* :ref:`Isolating Extension Modules <isolating-extensions-howto>`
+* :ref:`The Python 2.3 Method Resolution Order <python_2.3_mro>`
+* :ref:`Socket Programming HOWTO <socket-howto>`
+* :ref:`Timer file descriptor <timerfd-howto>`
+
+Debugging and profiling:
+
+* :ref:`Debugging C API extensions and CPython Internals with GDB <gdb>`
+* :ref:`Instrumenting CPython with DTrace and SystemTap <instrumentation>`
+* :ref:`Python support for the Linux perf profiler <perf_profiling>`
```
Comment (nedbat) [inline Doc/howto/index.rst]: You can use just the id, and Sphinx will find the title on the page and use it for the link:

```suggestion
General:

* :ref:`annotations-howto`
* :ref:`argparse-tutorial`
* :ref:`descriptorhowto`
* :ref:`enum-howto`
* :ref:`functional-howto`
* :ref:`ipaddress-howto`
* :ref:`logging-howto`
* :ref:`logging-cookbook`
* :ref:`regex-howto`
* :ref:`sortinghowto`
* :ref:`unicode-howto`
* :ref:`urllib-howto`

Advanced development:

* :ref:`curses-howto`
* :ref:`isolating-extensions-howto`
* :ref:`python_2.3_mro`
* :ref:`socket-howto`
* :ref:`timerfd-howto`

Debugging and profiling:

* :ref:`gdb`
* :ref:`instrumentation`
* :ref:`perf_profiling`
```

**Review by nedbat — COMMENTED** [review]
I love the re-org.  I suggested a less duplicative way to make .rst links.

**Comment by gvanrossum** [comment]
Would it make sense to distinguish between Python howto’s and C (extension) howto’s?

**Comment by willingc** [comment]
> Would it make sense to distinguish between Python howto’s and C (extension) howto’s?

Sure! Let me take a look after I am home.

**Comment by willingc** [comment]
> I love the re-org. I suggested a less duplicative way to make .rst links.

Eric Mathes and I were discussing this so he gets a bunch of credit for inspiration.

**Review by rhettinger — APPROVED** [review]

**Review by hugovk — APPROVED** [review]
👍 

And backport to 3.12?

File: Doc/howto/index.rst [hunk 2]
```diff
@@ -35,3 +33,31 @@ Currently, the HOWTOs are:
...
+* :ref:`unicode-howto`
+* :ref:`urllib-howto`
+
+Advanced development:
+
+* :ref:`curses-howto`
+* :ref:`isolating-extensions-howto`
+* :ref:`python_2.3_mro`
+* :ref:`socket-howto`
+* :ref:`timerfd-howto`
```
Comment (encukou) [inline Doc/howto/index.rst:57]: “Porting Extension Modules to Python 3” is a stub nowadays, but I don't think it should be removed in a reorganization PR.
```suggestion
* :ref:`timerfd-howto`
* :ref:`cporting-howto`
```

**Comment by encukou** [comment]
@willingc Is there more review/discussion you're waiting for, or should I merge this?

**Review by encukou — APPROVED** [review]

**Comment by encukou** [comment]
I'll merge.

**Comment by hugovk** [comment]
Backported to 3.12 to fix https://github.com/python/cpython/issues/127578.

**Pull request merged**